### PR TITLE
Define Plugin Status types

### LINKED ISSUE #1322: Defining Plugin Status
Not all plugins in the Plugin Catalog are as mature as others. Based on user discussion, I want to offer 3 status indicators that can be indicated in each plugin repository. 

**Updated Goal (1/17/17)** - provide support levels for plugins based on their maturity. 

--- 
~It should be clear if a plugin is known to run, known to *not* run or is somewhere in between.~

~Here they are:~
~**Stable** - This plugin is known to compile and run in the Snap framework. If you have any issues, please open an Issue.~
~**Not Working** - We’re not sure if this compiles or runs in the Snap framework. If you want to use it, please pick up open Issues and help it run again!~
~**Under Construction** - Master *should* compile and run in the Snap framework, but it may not work how you expect. See open Issues before using.~

### PR DESCRIPTION
**Pull request by mjbrender**
Fixes #1322

Summary of changes:
- New concept of plugins having a tiered status
- Created `docs/PLUGIN_STATUS.md` to outline matrix
- Reference statuses in the glossary & plugin authoring 
- Drafted an example of new columns for the Plugin Catalog: https://gist.github.com/mjbrender/7424d0f4edb29e7cc8030f4bdb6d05d9

Testing done:
- Reviewed with @bjray for commitment on statuses 
- Discussed with @nanliu on edits to plugin review process

@intelsdi-x/snap-maintainers

### CODE REVIEW

File: docs/PLUGIN_STATUS.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+## Approved Plugins
+
+These are primarily community-contributed plugins that meet or exceed the project's best practices. These plugins have been vetted by Snap maintainers as of the date listed in the Plugin Catalog. They are excellent references and are quite likely ready for use in your own Snap deployment.
+
+We prefer to not have other repositories that overlap with Approved plugins and suggest contributing to the existing version to help keep this list small and effective. Don't be shy about reaching out to existing plugin authors to see if you can help improve upon it.
+
+## Experimental
+
+These are plugins being developed that are not yet complete, but are shared with the community for feedback and testing. Think of this tier as an incubation phase that should move toward Approved or drop down to Unlabeled.
```
Comment (kjlyon) [inline docs/PLUGIN_STATUS.md]: Suggested rewording: ```These plugins are in development and are not yet complete. They are shared with the community for feedback and testing. Think of this tier... ```

File: docs/PLUGIN_STATUS.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+| Includes dependency management |                    |                    | :white_check_mark: | :white_check_mark: |
+| Includes CI status             |                    |                    | :white_check_mark: | :white_check_mark: |
+| Includes license               |                    |                    | :white_check_mark: | :white_check_mark: |
+| Support provided by a company  |                    |                    |                    | :white_check_mark: |
+
+## Supported Plugins
+
+These are our premier plugins for the Snap telemetry framework and its users. These are designed to follow all of our recommended practices. Issues are also closely monitored by its supporting company ([read more about that here]()). While companies may support some plugins they contribute to Snap, not all contributions will meet the Supported standard.
+
+We prefer to not have other repositories that overlap with Supported plugins and suggest contributing to the existing version to help keep this list small and effective. Like all plugins, community contribution is welcome.
```
Comment (kjlyon) [inline docs/PLUGIN_STATUS.md]: This paragraph is in _all_ of the sections below. Is there a better place above or below the description section where we can say this information only once.

File: docs/PLUGIN_STATUS.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+## Changing Stages 
+
+If you find a plugin that should move between stages (ex. from Unlabeled to Approved or Approved to Experimental), open an issue. 
+
+* Is it clear you can improve an existing plugin or 
+
+### What's the dill? 
+
+Snap is an open source project
```
Comment (kjlyon) [inline docs/PLUGIN_STATUS.md]: This section could use some rewording, suggestion:
```Snap is an open source project originated and actively supported by Intel. As other companies adopt Snap as their standard telemetry framework they may also choose to support  various plugins. We want to make sure you know which plugins and which companies are here to support you.```

File: docs/PLUGIN_AUTHORING.md [hunk 4]
```diff
@@ -127,7 +127,7 @@ In the plugin repo root directory, the `metadata.yml` file provides Snap project
 * **description**: paragraph describing the plugin's purpose
 * **badge**: a list of [badges](https://shields.io/) to display
 * **ci**: a list of ci services running for this repo
-* **status**: one of the three statuses [described below](#plugin-status)
+* **status**: one of the four statuses [described below](#plugin-status)
```
Comment (kjlyon) [inline docs/PLUGIN_AUTHORING.md]: This implies that the four statuses are described below. When in reality the four statuses are described in Plugin_Status.md

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -82,17 +83,26 @@ Your pull request should be rebased against the current master branch. Please do
 the current master branch in with your topic branch, nor use the Update Branch button provided
 by GitHub on the pull request page.
 
+## Submitting a Plugin
+
+Sharing a new plugin is one of the most impressive ways to jump into the Snap community as a contributor. This process is as simple as adding your plugin information to the [list in plugis.yml](docs/plugin.yml).
```
Comment (kjlyon) [inline CONTRIBUTING.md]: This section should link to Plugin_Authoring.md for description of how to and requirements.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -82,17 +83,26 @@ Your pull request should be rebased against the current master branch. Please do
...
+
+All plugins mature at their own pace and have varying levels of support from maintainers. Each status and its corresponding requirements are available in the [Plugin Status documentation](docs/PLUGIN_STATUS.md).
```
Comment (kjlyon) [inline CONTRIBUTING.md]: I don't think this is the right place to mention plugin status. I see it as something people must do for their plugin to be considered Supported. They will see this requirement and others when they go to Plugin_Authoring.md.

File: docs/PLUGIN_STATUS.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+These are plugins being developed that are not yet complete, but are shared with the community for feedback and testing. Think of this tier as an incubation phase that should move toward Approved or drop down to Unlabeled.
+
+We prefer to not have other repositories that overlap with Experimental plugins and suggest contributing to the existing version to help keep this list small and effective. Don't be shy about reaching out to existing plugin authors to see if you can help improve upon it.
+
+
+## All Other Plugins (Unlabeled)
+
+These are plugins in varying stages of completeness and are shared for reference. They do not necessarily follow best practices for plugin development. We welcome anyone forking these plugins and working toward Approved status.
+
+## Changing Stages
```
Comment (IRCody) [inline docs/PLUGIN_STATUS.md]: Should stages be changed to tiers?

File: docs/PLUGIN_STATUS.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+### What this is not
+
+* Not a definition of the plugin version - that's done through `releases`
+* Not a contract of support - we do our best as maintainers of Snap, but make no promises
+
+### Status Tiers
+
+Plugins will fall into one of the four tiers. Here are their brief overviews:
+* **Supported** - Designed by Intel with the intent of supporting customers
```
Comment (IRCody) [inline docs/PLUGIN_STATUS.md]: Designed by companies*?

File: docs/PLUGIN_STATUS.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+These are plugins in varying stages of completeness and are shared for reference. They do not necessarily follow best practices for plugin development. We welcome anyone forking these plugins and working toward Approved status.
+
+## Changing Stages 
+
+If you find a plugin that should move between stages (ex. from Unlabeled to Approved or Approved to Experimental), open an issue. 
+
+* Is it clear you can improve an existing plugin or 
+
+### What's the dill?
```
Comment (IRCody) [inline docs/PLUGIN_STATUS.md]: deal*

Comment (mjbrender) [inline docs/PLUGIN_STATUS.md]: You're putting me in a real pickle here @IRCody. Amiright? 

😬

**Comment by mjbrender** [comment]
Thanks so much for the reviews @kjlyon & @IRCody. I'll thumbs up issues as I resolve them. 

> What is the purpose of the Date Reviewed section? What value does that add?

The tiering is at pull request into the plugin catalog, but how can users know when it was last reviewed? That's why a date will be necessary.

File: docs/PLUGIN_STATUS.md (hunk 8 above)
Comment (mjbrender) [inline docs/PLUGIN_STATUS.md]: Linked it. The first one is a TL;DR version and connecting it to below is good.

File: docs/PLUGIN_STATUS.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+| Reasonable test coverage       |                    |                    | :white_check_mark: | :white_check_mark: |
+| Includes example tasks         |                    |                    | :white_check_mark: | :white_check_mark: |
+| Includes dependency management |                    |                    | :white_check_mark: | :white_check_mark: |
+| Includes CI status             |                    |                    | :white_check_mark: | :white_check_mark: |
+| Includes license               |                    |                    | :white_check_mark: | :white_check_mark: |
+| Support provided by a company  |                    |                    |                    | :white_check_mark: |
+
+## Supported Plugins
+
+These are our premier plugins for the Snap telemetry framework and its users. These are designed to follow all of our recommended practices. Issues are also closely monitored by its supporting company ([read more about that here]()). While companies may support some plugins they contribute to Snap, not all contributions will meet the Supported standard.
```
Comment (mjbrender) [inline docs/PLUGIN_STATUS.md]: Corrected - needed a better name for "What the dill" section before doing so :)

File: docs/PLUGIN_STATUS.md (hunk 2 above)
Comment (mjbrender) [inline docs/PLUGIN_STATUS.md]: I haven't thought of any. Do you have a recommendation @kjlyon? If not, it's worth repeating at the start.

**Comment by kjlyon** [comment]
It seems like our docs folder has many documents that are specifically about plugins. Inside the docs folder do you think we should create a Plugin folder?

**Comment by kjlyon** [comment]
Once a plugin is marked as Supported will people care when the date reviewed was?

File: CONTRIBUTING.md (hunk 5 above)
Comment (mjbrender) [inline CONTRIBUTING.md]: The next sentence does that.

File: CONTRIBUTING.md (hunk 6 above)
Comment (mjbrender) [inline CONTRIBUTING.md]: Great call - edited.

**Comment by mjbrender** [comment]
> It seems like our docs folder has many documents that are specifically about plugins. Inside the docs folder do you think we should create a Plugin folder?

We're avoiding that. Structure should really come from a tool that organizes the docs. See #1390

> Once a plugin is marked as Supported will people care when the date reviewed was?

Most definitely. Think about when we make big changes to the API or client libraries and expect people to update them. Maintainers can quickly look at Supported plugins updated before the change and find who needs to be notified about updates. Also, but not required, this update process can be visible to users through updates to "Date Reviewed" metadata via PR.

**Comment by kjlyon** [comment]
^^ Makes sense.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -82,17 +89,24 @@ Your pull request should be rebased against the current master branch. Please do
 the current master branch in with your topic branch, nor use the Update Branch button provided
 by GitHub on the pull request page.
 
+## Submitting a Plugin
+
+Sharing a new plugin is one of the most impressive ways to jump into the Snap community as a contributor. This process is as simple as adding your plugin information to the [list in plugis.yml](docs/plugin.yml). Read more about what is required of plugin authors in [Plugin Authoring](docs/PLUGIN_AUTHORING.md).
```
Comment (bjray) [inline CONTRIBUTING.md]: Do you mean to say "most impressive" or perhaps something like "easiest"?

File: docs/PLUGIN_STATUS.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
+# Plugin Status
+
+Not all plugins are designed with the same level of readiness in mind. To visualize that fact while also allowing everyone to share their work, we have introduced the concept of "tiers" of plugins. These tiers are communicated at statuses which will be visualized everywhere you see the [Plugin Catalog](PLUGIN_CATALOG.md).
```
Comment (bjray) [inline docs/PLUGIN_STATUS.md]: "communicated as statuses"

File: docs/PLUGIN_STATUS.md [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+## Approved Plugins
+
+These are primarily community-contributed plugins that meet or exceed the project's best practices. These plugins have been vetted by Snap maintainers as of the date listed in the Plugin Catalog. They are excellent references and are quite likely ready for use in your own Snap deployment.
+
+We prefer to not have other repositories that overlap with Approved plugins and suggest contributing to the existing version to help keep this list small and effective. Don't be shy about reaching out to existing plugin authors to see if you can help improve upon it.
+
+## Experimental
+
+These plugins are in development and are not yet complete, but are shared with the community for feedback and testing. They are shared with teh community for feedback and testing. Think of this tier as an incubation phase that should move toward Approved or drop down to Unlabeled.
```
Comment (bjray) [inline docs/PLUGIN_STATUS.md]: "with the community"

File: CONTRIBUTING.md (hunk 11 above)
Comment (mjbrender) [inline CONTRIBUTING.md]: I mean "most impressive" as a way to prompt people to recognize our appreciation for plugins. Plugin design is pretty straightforward, but it's by far not the easiest way to get involved. That would be through opening issues, tweeting about using it, blogging, etc.

**Comment by mjbrender** [comment]
Good eye @bjray. Corrected those typos - please update your review if it's good now.

File: CONTRIBUTING.md (hunk 11 above)
Comment (bjray) [inline CONTRIBUTING.md]: sounds good

**Comment by bjray** [comment]
LGTM

**Pull request merged**